### PR TITLE
Propagate early characteristic changes

### LINKED ISSUE #105: Early characteristic changes can be missed
As was seen in https://github.com/JuulLabs/kable/issues/99#issuecomment-842941308, an unprovoked (prior to writing to the config descriptor to enable Notify/Indicate) characteristic may trigger a "characteristic change" event.

Code used in test branch that should be brought in to main code:

<details>
<summary><code>Observers.kt</code></summary>

```kotlin
    fun acquire(characteristic: Characteristic) = flow {
        try {
            characteristicChanges
                .onSubscription {
                    peripheral.suspendUntilReady()

                    if (observers.incrementAndGet(characteristic) == 1) {
                        peripheral.startObservation(characteristic)
                    }
                }
                .collect {
                    if (it.characteristic.characteristicUuid == characteristic.characteristicUuid &&
                        it.characteristic.serviceUuid == characteristic.serviceUuid
                    ) emit(it.data)
                }
        } finally {
            if (observers.decrementAndGet(characteristic) < 1) {
                try {
                    peripheral.stopObservation(characteristic)
                } catch (e: NotReadyException) {
                    // Silently ignore as it is assumed that failure is due to connection drop, in which case Android
                    // will clear the notifications.
                    Log.d(TAG, "Stop notification failure ignored.")
                }
            }
        }
    }
```
</details>

### LINKED ISSUE #112: peripheral.observe(readCharacteristic) doesn't provide a form to know when is safe to write
Hello there!

I'm trying to migrate a project to kable. The ble device (and others I worked with...) uses some opcode communication where you write to a writeCharacteristic and all the data is received on the readCharacteristic subscribed to notify.
After notify is subscribed I can enqueue calls to get a first response as soon as posible.

Enter kable, the peripheral.observe(readCharacteristic) provides a flow with the data, but cant signal when it is actually able to receive data. So i can't just enqueue a write to the descriptor as doing it "on the next line" because the first response will be missed as the observe isn't really ready. I can workaround that doing a delay(2000) for example between the .observe and the write.

Maybe i've missed something but can't seem to find a method or param that handles that.
Could it be possible to set the notify/observe on the new (0.6) connect params? or maybe provide a param to the observe to set an initial write?

Keep the good work, awesome library!

### PR DESCRIPTION
**Pull request by twyatt**
Closes #105
Closes #112

Also fixes JavaScript observations to wait until connection is ready. Only users of experimental advertisement support in Chrome would've seen this bug, as using `requestDevice` obtains devices in a connected state, i.e. already ready.

### CODE REVIEW

File: core/src/androidMain/kotlin/Observers.kt [hunk 1]
```diff
@@ -38,20 +40,21 @@ internal class Observers(
...
-            peripheral.startObservation(characteristic)
-        }
-
-        try {
-            characteristicChanges.collect {
-                if (it.characteristic.characteristicUuid == characteristic.characteristicUuid &&
-                    it.characteristic.serviceUuid == characteristic.serviceUuid
-                ) emit(it.data)
+    fun acquire(
+        characteristic: Characteristic
```
Comment (sdonn3) [inline core/src/androidMain/kotlin/Observers.kt:44]: nit: seems like a single parameter could just go on one line?

**Review by sdonn3 — APPROVED** [review]
looks good!

**Review by cedrickcooke — APPROVED** [review]

**Comment by twyatt** [comment]
@sdonn3 @cedrickcooke re-requested review as I've added a additional changes since last reviewed:

- refactored JavaScript observation handling to resemble the other platforms
- implemented `observe`'s `onSubscription` lambda to be invoked on re-connection
    - now supports invoking all `onSubscription` lambdas when multiple `observe` Flows are created for the same `characteristic`

File: core/src/commonMain/kotlin/Peripheral.kt [hunk 2]
```diff
@@ -114,8 +115,27 @@ public interface Peripheral {
...
+     * The [onSubscription] action is useful in situations where an initial operation is needed when starting an
+     * observation (such as writing a configuration to the peripheral and expecting the response to come back in the
+     * form of a characteristic change). The [onSubscription] is invoked for every new subscriber; if it is desirable to
+     * only invoke the [onSubscription] once per connection (for the specified [characteristic]) then you can either
+     * use the [shareIn][kotlinx.coroutines.flow.shareIn] [Flow] operator on the returned [Flow], or call [observe]
+     * again with the same [characteristic] and without specifying an [onSubscription] action.
+     *
+     * If multiple [observations][observe] are created for the same [characteristic] but with different [onSubscription]
+     * actions, then the [onSubscription] actions will be executed in the order in which the returned [Flow]s are
+     * subscribed to.
```
Comment (cedrickcooke) [inline core/src/commonMain/kotlin/Peripheral.kt:135]: ❤️

File: core/src/jsMain/kotlin/Observers.kt [hunk 3]
```diff
@@ -1,129 +1,89 @@
...
+            peripheral.startObservation(characteristic)
+            onSubscriptionActions.forEach { it() }
         }
     }
+}
 
-    fun invalidate() {
-        observers.forEach { (_, observation) ->
-            observation.listener = null
+private class Observations {
```
Comment (cedrickcooke) [inline core/src/jsMain/kotlin/Observers.kt:52]: It's a little sad that this `Observations` class can't live in common code. I can see that they're different but it feels _super_ similar between the three implementations.

**Review by cedrickcooke — APPROVED** [review]

File: core/src/jsMain/kotlin/Observers.kt (hunk 3 above)
Comment (twyatt) [reply core/src/jsMain/kotlin/Observers.kt:52]: I agree it's a little sad. Maybe one day I can figure out a way to make it generic and reusable, but for now, just getting it working. 🤷

**Comment by twyatt** [comment]
Going to do some internal testing against the JavaScript target, then merge.

File: core/src/androidMain/kotlin/Observers.kt [hunk 4]
```diff
@@ -34,25 +36,26 @@ internal class Observers(
...
-                if (it.characteristic.characteristicUuid == characteristic.characteristicUuid &&
-                    it.characteristic.serviceUuid == characteristic.serviceUuid
-                ) emit(it.data)
+    fun acquire(
+        characteristic: Characteristic,
+        onSubscription: OnSubscriptionAction,
+    ) = characteristicChanges
+        .onSubscription {
+            peripheral.suspendUntilReady()
+            if (observations.add(characteristic, onSubscription) == 1) {
```
Comment (sdonn3) [inline core/src/androidMain/kotlin/Observers.kt:47]: Should this be '>= 1' to mirror the '< 1' below?

**Review by sdonn3 — APPROVED** [review]
Totally appreciate how easy this is to read through

**Comment by twyatt** [comment]
lol, @sdonn3 either I'm going crazy, or you left a comment then deleted it.

Here was the response I was going to do, though I'm guessing you don't need it now. 🤷 

----

Technically, the `< 1` below could be a `0` (but `< 1` was a little bit of a safety buffer), whereas this can only be a `== 1`.

I'll explain why:

In `onSubscription` we "register", aka "add", a subscription to the `Map` of characteristics to subscriptions. The return is the number of subscribers we have _after_ "registering" (or aka "adding"). The underlying system only needs to setup BLE notifications/indications once per characteristic (and not more than once). So, we essentially only want to enable them when we go from `0` to `1` subscribers and we'll just leave the notifications/indications enabled as more subscribers jump on.

On the reverse side (in `onCompletion`) we need to keep track as `Flow` collectors unsubscribe, all the way down until we have no more subscribers, and can therefore stop the BLE notifications/indications for that characteristic. Since we're subtracting within the `observations.remove` I thought it were safer if I messed up the math (and we somehow ended up with `-1`) that checking for `< 1` would be good. But, assuming the maths checks out 😉  then `== 0` would've sufficed. Essentially it would've represented when we are transitioning from `1` to `0` subscribers.

**Pull request merged**